generate-lockfiles --sync fails with PEX backend when lockfile already exists

**Describe the bug**
Updating existing lockfiles with `pants generate-lockfiles --sync` fails.
This is caused by the metadata comments not getting stripped from the digest contents in `generate_pex_lockfiles` before being passed to PEX.

The bug can be avoided by adding the `separate_lockfile_metadata_file = true` option to the `pants.toml` file, as mentioned in [this blog post](https://www.pantsbuild.org/blog/2025/12/11/lockfile-metadata-files).

Steps to reproduce:

1. Generate a lockfile
  ```shell
  $ pants generate-lockfiles --resolve=python-default 
  11:56:06.77 [INFO] Completed: Generate pex lockfile for python-default
  11:56:06.80 [INFO] Wrote lockfile for the resolve `python-default` to 3rdparty/python/default_lock.txt
  ```

2. Update the lockfile with the sync option
  ```shell
  $ pants generate-lockfiles --resolve=python-default --sync
  11:54:07.43 [INFO] Canceled: Generate pex lockfile for python-default
  11:54:09.22 [INFO] Completed: Generate pex lockfile for python-default
  11:54:09.24 [ERROR] 1 Exception encountered:
  
  Engine traceback:
    in root
      ..
    in pants.core.goals.generate_lockfiles.generate_lockfiles_goal
      `generate-lockfiles` goal
  
  Traceback (most recent call last):
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/core/goals/generate_lockfiles.py", line 593, in generate_lockfiles_goal
      results = await concurrently(
                ^^^^^^^^^^^^^^^^^^^
      ...<9 lines>...
      )
      ^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/internals/selectors.py", line 238, in Concurrently
      return await _Concurrently(tuple(__arg0))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/internals/selectors.py", line 49, in __await__
      result = yield self.calls
               ^^^^^^^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/rules.py", line 66, in wrapper
      return await call
             ^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/internals/selectors.py", line 37, in __await__
      result = yield self
               ^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/backend/python/goals/lockfile.py", line 196, in generate_pex_lockfile
      result = await execute_process_or_raise(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      ...<56 lines>...
      )
      ^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/rules.py", line 66, in wrapper
      return await call
             ^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/internals/selectors.py", line 37, in __await__
      result = yield self
               ^^^^^^^^^^
    File "/nasfs/users/u0167447/.cache/nce/4f07959fafe3296d25d3e3d2c92165cc6933455f0caa80a335a8aaa14c295fb3/bindings/venvs/2.32.0/lib/python3.14/site-packages/pants/engine/process.py", line 390, in fallible_to_exec_result_or_raise
      raise ProcessExecutionFailure(
      ...<5 lines>...
      )
  pants.engine.process.ProcessExecutionFailure: Process 'Generate pex lockfile for python-default' failed with exit code 1.
  stdout:
  
  stderr:
  The lock file at 3rdparty/python/default_lock.txt does not contain valid JSON: Expecting value: line 1 column 1 (char 0)
  
  
  
  Use `--keep-sandboxes=on_failure` to preserve the process chroot for inspection.
  ```

**Pants version**
2.32.0

**OS**
Linux
Kernel 6.12.0-124.55.3.el10_1.x86_64

--- a/docs/notes/2.33.x.md
+++ b/docs/notes/2.33.x.md
@@ -62,6 +62,8 @@ Fixed a bug in Ruff backend where ancestor `__init__.py` files were not included
 
 All version of [Ruff](https://docs.astral.sh/ruff/) from [0.15.6](https://github.com/astral-sh/ruff/releases/tag/0.15.14) to [0.15.14](https://github.com/astral-sh/ruff/releases/tag/0.15.14) are now included in `default_known_versions`.
 
+Fixed a bug where `--sync`-ing a Pex lockfile with legacy metadata headers failed.
+
 #### Protobuf
 
 Upgraded the default version of `buf` to v1.69.0 (previously v1.3.0).

--- a/src/python/pants/backend/python/goals/lockfile.py
+++ b/src/python/pants/backend/python/goals/lockfile.py
@@ -32,6 +32,7 @@ from pants.backend.python.util_rules.pex_requirements import (
     ResolveConfig,
     ResolveConfigRequest,
     determine_resolve_config,
+    strip_comments_from_pex_json_lockfile,
 )
 from pants.backend.python.util_rules.uv import UvEnvironment, generate_pyproject_toml
 from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
@@ -189,6 +190,16 @@ async def generate_pex_lockfile(
                 conjunction=GlobExpansionConjunction.any_match,
             )
         )
+
+        if existing_lockfile_digest != EMPTY_DIGEST:
+            existing_lockfile_contents = await get_digest_contents(existing_lockfile_digest)
+            if existing_lockfile_contents:
+                stripped = strip_comments_from_pex_json_lockfile(
+                    existing_lockfile_contents[0].content
+                )
+                existing_lockfile_digest = await create_digest(
+                    CreateDigest([FileContent(req.lockfile_dest, stripped)])
+                )
     else:
         existing_lockfile_digest = EMPTY_DIGEST
 

--- a/src/python/pants/backend/python/goals/lockfile_test.py
+++ b/src/python/pants/backend/python/goals/lockfile_test.py
@@ -404,6 +404,69 @@ def test_find_links_scoped_to_resolve() -> None:
     assert result_by_resolve["b"].find_links == FrozenOrderedSet([])
 
 
+def test_sync_strips_embedded_metadata_header(rule_runner: PythonRuleRunner) -> None:
+    """Regression test for https://github.com/pantsbuild/pants/issues/23398."""
+    rule_runner.set_options(
+        [
+            "--python-resolves={'test': 'test.lock'}",
+            "--python-resolver=pex",
+            "--no-python-separate-lockfile-metadata-file",
+        ],
+        env_inherit=PYTHON_BOOTSTRAP_ENV,
+    )
+
+    result = rule_runner.request(
+        GenerateLockfileResult,
+        [
+            GeneratePexLockfile(
+                requirements=FrozenOrderedSet(["ansicolors==1.1.8"]),
+                find_links=FrozenOrderedSet([]),
+                interpreter_constraints=InterpreterConstraints(),
+                resolve_name="test",
+                lockfile_dest="test.lock",
+                diff=False,
+                lock_style="universal",
+                complete_platforms=(),
+            )
+        ],
+    )
+    rule_runner.write_digest(result.digest)
+
+    rule_runner.set_options(
+        [
+            "--python-resolves={'test': 'test.lock'}",
+            "--python-resolver=pex",
+            "--no-python-separate-lockfile-metadata-file",
+            "--generate-lockfiles-sync",
+        ],
+        env_inherit=PYTHON_BOOTSTRAP_ENV,
+    )
+    sync_result = rule_runner.request(
+        GenerateLockfileResult,
+        [
+            GeneratePexLockfile(
+                requirements=FrozenOrderedSet(["ansicolors==1.1.8"]),
+                find_links=FrozenOrderedSet([]),
+                interpreter_constraints=InterpreterConstraints(),
+                resolve_name="test",
+                lockfile_dest="test.lock",
+                diff=False,
+                lock_style="universal",
+                complete_platforms=(),
+            )
+        ],
+    )
+    sync_digest_contents = rule_runner.request(DigestContents, [sync_result.digest])
+    assert len(sync_digest_contents) == 1
+    content = sync_digest_contents[0].content.decode()
+    header, _, body = content.partition("\n\n")
+    assert header.strip() != ""
+    lock_content = json.loads(body)
+    reqs = lock_content["locked_resolves"][0]["locked_requirements"]
+    assert len(reqs) == 1
+    assert reqs[0]["project_name"] == "ansicolors"
+
+
 def test_empty_requirements(rule_runner: PythonRuleRunner) -> None:
     with pytest.raises(ExecutionError) as excinfo:
         json.loads(

